### PR TITLE
Add product and product query

### LINKED ISSUE #2: Add new entity products
Products are all the things available. For example, tickets, memberships, partnerships, shirts, goods

- [x] Product type is a union of types
  - [x] Ticket
  - [x] Membership
  - [x] Partnership
  - [ ] Food
  - [ ] Shirt (?)
  - others tbd
- [x] The product types are based around an interface for common required fields
- [x] CRUD product
- [x] Query product
- [x] PagedProduct type

### PR DESCRIPTION
**Pull request by brettski**
Add products
some restructuring of the overall project

resolves #2

### CODE REVIEW

**Comment by brettski** [comment]
All the pieces for products are there. probably needs a bit of clean up here and there.

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "that-api-garage",
   "version": "1.0.0",
-  "description": "THAT Garage api",
+  "description": "THAT Garage api. Where all the other stuff goes",
   "main": "index.js",
   "engines": {
     "node": "14.x"
```
Comment (csell5) [inline package.json:7]: Are we really running 14 on the containers?

File: src/index.js [hunk 2]
```diff
@@ -103,8 +92,7 @@ function createUserContext(req, res, next) {
     locale: req.headers.locale,
     authToken: req.headers.authorization,
     correlationId,
-    sentry: Sentry,
-    enableMocking: enableMocking(),
+    enableMocking: false,
```
Comment (csell5) [inline src/index.js:95]: I take it we're using this elsewhere (that-api?)? if so, I'd set a better default there, and drop it here.

File: src/lib/__tests__/productDateForge.test.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Required for @thatconference/api
+process.env.INFLUX_TOKEN = 'TEST_INFLUX_TOKEN_VALUE';
```
Comment (csell5) [inline src/lib/__tests__/productDateForge.test.js:2]: You should really set these up as part of the test setup/teardown such that it doesn't bleed into other testing and returns it back to what it was when it started.

File: src/graphql/typeDefs/mutations/productCreate.graphql [hunk 4]
```diff
@@ -0,0 +1,13 @@
+"create a product by type of product"
+type ProductCreate {
+  "create a ticket type product"
+  ticket(ticket: TicketCreateInput!): Ticket! @auth(requires: "admin")
+
+  "create a membership type product"
+  membership(membership: MembershipCreateInput!): Membership!
+    @auth(requires: "adming")
```
Comment (csell5) [inline src/graphql/typeDefs/mutations/productCreate.graphql:8]: role spelled wrong.

**Review by csell5 — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (brettski) [reply package.json:7]: Nope, but it is the plan. Knock it back to 12 for now?

File: src/index.js (hunk 2 above)
Comment (brettski) [reply src/index.js:95]: Correct, Sentry is not used from the context so it was removed.

File: package.json (hunk 1 above)
Comment (csell5) [reply package.json:7]: I wasn't sure if the build/container was reading that to do what it does. if it's not, then I wouldn't worry about it.,

File: src/index.js (hunk 2 above)
Comment (csell5) [reply src/index.js:95]: I actually wasn't talking about senty.. but rather the enable mocking.

File: src/lib/__tests__/productDateForge.test.js (hunk 3 above)
Comment (brettski) [reply src/lib/__tests__/productDateForge.test.js:2]: Added a afterAll() function to remove process vars.

File: src/graphql/typeDefs/mutations/productCreate.graphql (hunk 4 above)
Comment (brettski) [reply src/graphql/typeDefs/mutations/productCreate.graphql]: fixed. And yes I need to add create, update and get requests for partnership and membership which would have caught this.

File: src/lib/__tests__/productDateForge.test.js (hunk 3 above)
Comment (csell5) [reply src/lib/__tests__/productDateForge.test.js:2]: but that assumes they wen't actually there before. I would do something like this

```
describe ...
 const OLD_ENV = process.env; // save for later

  beforeEach(() => {
    jest.resetModules() // clears the cache
    process.env = { ...OLD_ENV }; // makes a copy
   process.env.THING = 'THANG'; // sets the new stuff
  });

  afterAll(() => {
    process.env = OLD_ENV; // sets it back to what it was
  });
```

Comment (brettski) [reply src/lib/__tests__/productDateForge.test.js:2]: Can't do that on before each as the are required for ` @thatconference/api` to load. I tried using a beforeAll(), but it complained that the variable wasn't a function when trying to run the tests.

Comment (csell5) [reply src/lib/__tests__/productDateForge.test.js:2]: I see.. this is why you're using require and doing it after the process.env is set.. that's some internal weirdness that would could easily get refactored by a change in prettier. I'd suggest. Pulling that require into the before all, then you can control all of that, including when it get's pulled in.

**Review by csell5 — APPROVED** [review]

**Pull request merged**